### PR TITLE
post renewal hook: Add RENEWED_DOMAINS and FAILED_DOMAINS as environment variables

### PR DESCRIPTION
**Pull request by mcassaniti**
This change addresses #9723. Since it was relatively trivial to implement I went ahead and wrote a PR for review.

## Pull Request Checklist

- [x] The Certbot team has recently expressed interest in reviewing a PR for this. If not, this PR may be closed due our limited resources and need to prioritize how we spend them.
- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Add or update any documentation as needed to support the changes in this PR.
- [ ] Include your name in `AUTHORS.md` if you like.

### CODE REVIEW

**Comment by mcassaniti** [comment]
I couldn't get the tests to run locally because tox was failing to setup the environment properly. I'll need a bit of guidance.

**Comment by bmw** [comment]
Nice! I see the appeal of this and it's a clean and straightforward enough change that I think I'd be happy to review and accept it.

What problems are you having setting up the local environment? Have you seen https://certbot.eff.org/docs/contributing.html#getting-started?

**Comment by mcassaniti** [comment]
RTFM helped fix the tox issue. Without it the dependencies that tox wants to install failed.

File: certbot/docs/using.rst
Comment (bmw) [inline certbot/docs/using.rst:710]: We can wait right before merging to change this, but I think we should mention the version number where this feature became available. There's a lot of old Certbots entombed in various distros out there.

**Review by bmw — COMMENTED** [review]
In addition to my documentation nitpick, I have three higher level concerns about this after digging in more closely:

1. I think we should probably modify `certbot.hooks.post_hook` to accept domains lists instead or as well. Otherwise, post hooks will get these environment variables when using `certbot renew` but not with other subcommands.
2. I worry a bit about the list of domains becoming too long to shove in the OS environment. Imagine you're a shared hosting provider for many thousands of domains and many of these certificates try to renew at once. Any ideas on how we can eliminate/reduce the potential of breaking Certbot in this scenario?
3. I think we should add a simple unit or integration test for this feature so we don't accidentally break it in the future. Integration tests can be found under the top level `certbot-ci` directory.

Let me know what you think!

File: certbot/docs/using.rst
Comment (mcassaniti) [reply certbot/docs/using.rst:710]: I ran off the checklist and assumed I'd need to update all the docs. I can revert if you wish?

**Comment by mcassaniti** [comment]
> In addition to my documentation nitpick, I have three higher level concerns about this after digging in more closely:
> 
> 1. I think we should probably modify `certbot.hooks.post_hook` to accept domains lists instead or as well. Otherwise, post hooks will get these environment variables when using `certbot renew` but not with other subcommands.

I've updated the code to include this.

> 
> 2. I worry a bit about the list of domains becoming too long to shove in the OS environment. Imagine you're a shared hosting provider for many thousands of domains and many of these certificates try to renew at once. Any ideas on how we can eliminate/reduce the potential of breaking Certbot in this scenario?

Perhaps a hard coded limit on the string might do. Note that deploy hooks suffer the same problem [here](https://github.com/certbot/certbot/blob/4fc4d536c1acc65646189c5b5aaf02aaa88c3f4a/certbot/certbot/_internal/hooks.py#L227).

> 
> 3. I think we should add a simple unit or integration test for this feature so we don't accidentally break it in the future. Integration tests can be found under the top level `certbot-ci` directory.

I've added unit tests. Compared to `certbot.hooks.deploy_hook`, I haven't leaked OS environment variables.

> 
> 
> Let me know what you think!

File: certbot/docs/using.rst
Comment (bmw) [reply certbot/docs/using.rst:710]: This is a response to https://github.com/certbot/certbot/pull/9724/files#r1243102591 just in case GitHub's UI doesn't make that clear.

Sorry! I wasn't clear in my previous comment. I was trying to suggest that we add a phrase to the beginning of what you wrote here like "Starting with Certbot 2.7.0, ...". My thinking is that there are plenty of old versions of Certbot out there that people are using which will not have this behavior. 

I also said that we can wait until right before we merge this to make this change if you want to avoid having to update it again if this PR isn't included in our next release.

**Review by bmw — COMMENTED** [review]
> Perhaps a hard coded limit on the string might do. Note that deploy hooks suffer the same problem [here](https://github.com/certbot/certbot/blob/4fc4d536c1acc65646189c5b5aaf02aaa88c3f4a/certbot/certbot/_internal/hooks.py#L227).

I think that's a bit different because I think the existing variable will only be the domains from one certificate while the new variable will be the domains from all certificates that were attempted to be renewed. I agree a somewhat arbitrary but conservative limit on the string or total environment size is probably a good idea. I'd also add logging message when this variable isn't set or is truncated due to its size. I don't off the top of my head know what that limit should be across all major OSes though.

We could additionally switch from `{RENEWED,FAILED}_DOMAINS` to `{RENEWED,FAILED}_LINEAGES` if that'd still be sane for your use case, but even then, I think we should probably add a conservative limit to avoid this being a breaking change for people.

This PR otherwise LGTM.

File: certbot/docs/using.rst
Comment (mcassaniti) [reply certbot/docs/using.rst:710]: I'll take the risk that this is merged for 2.7.0 and update accordingly.

**Comment by mcassaniti** [comment]
After looking across Windows, Linux and FreeBSD I took a stab and went with a capped environment variable limit of around 32,000 characters. I was surprised there wasn't code limiting this within the CPython codebase already which I was hoping to copy.

File: certbot/certbot/_internal/hooks.py
Comment (bmw) [inline certbot/certbot/_internal/hooks.py:148]: This should fix the relevant test failures. Merging `master` into this branch may not be necessary, but it should definitely fix the unrelated ones.

```suggestion
                'RENEWED_DOMAINS': renewed_domains_str,
```

Comment (mcassaniti) [reply certbot/certbot/_internal/hooks.py]: Whoops. Fixed.

**Review by bmw — APPROVED** [review]
LGTM! Thanks mcassaniti.

**Pull request merged**
